service/history/historyReplicator.go

```diff
@@ -93,10 +101,43 @@ func (r *historyReplicator) ApplyEvents(request *h.ReplicateEventsRequest) (retE
 			return err
 		}
 
-		// TODO
-		// WARNING
-		// CODE SHOULD BE REMOVE WHEN IN PROD
-		// TEMP CODE LOGIC FOR TESTING ONLY
+		rState := msBuilder.replicationState
+		// Check if this is a stale event
+		if rState.CurrentVersion > request.GetVersion() {
+			// Replication state is already on a higher version, we can drop this event
+			// TODO: We need to replay external events like signal to the new version
+			r.logger.Warnf("Dropping stale replication task.  Current Version: %v, Task Version: %v", rState.CurrentVersion,
+				request.GetVersion())
+			return nil
+		}
+
+		// Check if this is the first event after failover
+		if rState.CurrentVersion < request.GetVersion() {
+			previousActiveCluster := r.metadataMgr.ClusterNameForFailoverVersion(rState.CurrentVersion)
```

Comment:
i think you should use LastWriteVersion instead of CurrentVersion

Reply:
Done.